lib/improver/metadata/set_attributes.py

```diff
@@ -0,0 +1,144 @@
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""
+Module setting attributes for IMPROVER blended data, nowcasts and spot
+forecasts
+"""
+
+import re
+import warnings
+
+from improver.metadata.constants.attributes import (
+    STANDARD_GRID_TITLE_STRING,
+    UK_SPOT_TITLE_STRING,
+    GLOBAL_SPOT_TITLE_STRING,
+    DATASET_ATTRIBUTES)
+from improver.utilities.cube_metadata import amend_metadata
+
+
+def set_product_attributes(cube, product):
+    """
+    Set attributes on an output cube of type matching a key string in the
+    DATASET_ATTRIBUTES dictionary.
+
+    Args:
+        cube (iris.cube.Cube):
+            Cube containing product data
+        product (str):
+            String describing product type
+
+    Returns:
+        updated_cube (iris.cube.Cube):
+            Cube with updated attributes
+    """
+    try:
+        original_title = cube.attributes["title"]
+    except KeyError:
+        original_title = ""
+
+    try:
+        updated_cube = amend_metadata(
+            cube, attributes=DATASET_ATTRIBUTES[product])
+    except KeyError:
+        options = list(DATASET_ATTRIBUTES.keys())
+        raise ValueError(
+            "product '{}' not available (options: {})".format(
+                product, options))
+
+    if STANDARD_GRID_TITLE_STRING in original_title:
+        updated_cube.attributes["title"] += " on {}".format(
+            STANDARD_GRID_TITLE_STRING)
+
+    return updated_cube
+
+
+def _match_title(original_title):
```

Comment:
From my perspective, it would be easier if there was a standalone unit test testing this function. This would make it much clearer what this function is actually doing itself, and would be helpful if I ever had to develop this code in the future.

Reply:
While I understand that it's easier to apply a blanket rule, I think this case is an exception covered by the "unless the logic is reasonably obvious" and "the code is tested elsewhere" guidance here: https://github.com/metoppv/improver/wiki/Code-Style-Guide#unit-testing

This function is a one-line regular expression match.  The only reason I have factored it out is so as to be able to call a single, self-explanatory line in `update_spot_metadata` in place of the relatively long line-wrapped regular expression, to increase the readability of the code.  Writing unit tests would be equivalent to duplicating unit tests for the in-built `re` library functionality, which adds code to maintain in our code base without adding value.

I've added further detail to the docstring to make it clearer what this function is doing, and am happy to further extend that if you feel it's still not clear.

Reply:
Yes, I think you're right that we don't want to be retesting functionality from the `re` library. My interest would be to just test that the `field` and `grid` groups are filled correctly. Following putting together the test below, for some reason it seems like `expected_grid` assertion fails because `result.group("grid")` is an empty string, which I didn't think was the intended behaviour. This might be worth looking into.

```python
class Test__match_title(unittest.TestCase):

    def test_basic(self):
        expected_field = "MOGREPS-UK Temperature Forecast"
        expected_grid = "UK 2 km Standard Grid"
        original_title = (
            "MOGREPS-UK Temperature Forecast on UK 2 km Standard Grid")
        result = _match_title(original_title)
        self.assertEqual(result.group("field"), expected_field)
        self.assertEqual(result.group("grid"), expected_grid)
```